### PR TITLE
feature(whl): add AWR algorithm.

### PR DESCRIPTION
**Pull request by kxzxvbk**
## Description
Implement the algorithm of AWR (languange model as the policy)

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [x] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/policy/prompt_awr.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+            # calculate PG loss
+            real_act = batch['action']  # shape: (B, shot_number)
+            if len(real_act.shape) == 1:
+                real_act = real_act.unsqueeze(-1)
+            # Calculate loss.
+            total_policy_loss, total_entropy_loss, total_value_loss = 0, 0, 0
+            for ii in range(self._cfg.shot_number):
+                log_prob = output['dist'].log_prob(real_act[:, ii])
+                policy_loss = -(log_prob * torch.exp((return_ - batch['value']) / self._cfg.learn.beta)).mean()
```
Comment (PaParaZz1) [inline ding/policy/prompt_awr.py:152]: maybe there is a `clamp` operation about the adv weight

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Comment by PaParaZz1** [comment]
add the AWR algorithm into the table of  README

File: ding/model/template/language_transformer.py [hunk 2]
```diff
@@ -36,10 +39,16 @@ def __init__(
...
             defaults to be ``True``.
+            - hidden_dim (:obj:`int`): The embedding dimension of the encoding model (e.g. BERT). This value should \
+            correspond to the model you use. For bert-base-uncased, this value is 768.
+            - norm_embedding (:obj:`bool`): Whether to normalize the embedding vectors. Default to be ``False``.
         """
         super().__init__()
         self.tokenizer = AutoTokenizer.from_pretrained(model_name)
         self.model = AutoModelForTokenClassification.from_pretrained(model_name)
+        in_channel = hidden_dim if not add_linear else embedding_size
+        self.value_head = nn.Linear(in_channel, 1)
```
Comment (puyuan1996) [inline ding/model/template/language_transformer.py:50]: maybe we can add a discrete regression option like this [paper](https://arxiv.org/abs/2403.03950)

File: ding/model/template/language_transformer.py [hunk 3]
```diff
@@ -18,13 +18,16 @@ class LanguageTransformer(nn.Module):
     Interfaces:
         ``__init__``, ``forward``
     """
+    mode = ['compute_actor', 'compute_critic', 'compute_actor_critic']
 
     def __init__(
             self,
             model_name: str = "bert-base-uncased",
```
Comment (puyuan1996) [inline ding/model/template/language_transformer.py:25]: add some comments about why we use "bert-base-uncased" as default?

File: ding/policy/prompt_awr.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        self._learn_model.train()
+
+        for i in range(0, len(data), self._cfg.learn.batch_size):
+            batch = default_collate(data[i:i + self._cfg.learn.batch_size])
+            if self._cuda:
+                batch = to_device(batch, self._device)
+
+            # Prepare train_sample (the question to be answered) and the candidate_samples (the prompts to be selected)
+            train_samples, cand_samples = batch["obs"]["train_sample"], batch["obs"]["candidate_samples"]
+            for ii in range(len(cand_samples)):
```
Comment (puyuan1996) [inline ding/policy/prompt_awr.py:144]: change 'ii' into more explainable name?

File: ding/policy/prompt_awr.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            if len(real_act.shape) == 1:
+                real_act = real_act.unsqueeze(-1)
+            # Calculate loss.
+            total_policy_loss, total_entropy_loss, total_value_loss = 0, 0, 0
+            for ii in range(self._cfg.shot_number):
+                log_prob = output['dist'].log_prob(real_act[:, ii])
+                adv = torch.clamp(
+                    return_ - batch['value'], min=self._cfg.learn.norm_range[0], max=self._cfg.learn.norm_range[1]
+                )
+                policy_loss = -(log_prob * torch.exp(adv / self._cfg.learn.beta)).mean()
```
Comment (puyuan1996) [inline ding/policy/prompt_awr.py:160]: add comments about the key part of advantage weighted regression

Comment (puyuan1996) [reply ding/policy/prompt_awr.py:160]: Should the operation `torch.exp(adv / self._cfg.learn.beta)` stop the gradient flow?

File: ding/policy/prompt_awr.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+                cand_samples[ii] = cand_samples[ii][0]
+            output = self._learn_model.forward(train_samples, cand_samples, mode='compute_actor_critic')
+            return_ = batch['return']
+
+            # calculate PG loss
+            real_act = batch['action']  # shape: (B, shot_number)
+            if len(real_act.shape) == 1:
+                real_act = real_act.unsqueeze(-1)
+            # Calculate loss.
+            total_policy_loss, total_entropy_loss, total_value_loss = 0, 0, 0
```
Comment (puyuan1996) [inline ding/policy/prompt_awr.py:154]: update comments

File: ding/policy/prompt_awr.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+            # (float) Term added to the denominator to improve numerical stability in optimizer.
+            eps=1e-8,
+            # (float) Maximum norm for gradients.
+            grad_norm=0.5,
+            # (float) Scaling factor for value network loss relative to policy network loss.
+            value_weight=0.5,
+            # (float) Coefficient that controls the exp scale in awr algorithm.
+            beta=1.0,
+            # (float) Weight of entropy regularization in the loss function.
+            entropy_weight=0.01,
```
Comment (puyuan1996) [inline ding/policy/prompt_awr.py:57]: Should we change it to a more generally applicable constant, such as 0.001?

File: ding/policy/prompt_awr.py (hunk 5 above)
Comment (PaParaZz1) [reply ding/policy/prompt_awr.py:160]: this operation is computed on the batch data rather than the output data, thus it is no need to stop the gradient

File: dizoo/tabmwp/config/tabmwp_awr_config.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
+from easydict import EasyDict
+
+tabmwp_prompt_pg_config = dict(
+    exp_name='tabmwp_prompt_pg_seed0',
```
Comment (PaParaZz1) [inline dizoo/tabmwp/config/tabmwp_awr_config.py:4]: polish name, not pg in this file

File: ding/model/template/language_transformer.py [hunk 9]
```diff
@@ -36,10 +39,16 @@ def __init__(
             - embedding_size (:obj:`int`): The embedding size of the added linear layer, such as 128.
             - freeze_encoder (:obj:`bool`): Whether to freeze the encoder language model while training, \
             defaults to be ``True``.
+            - hidden_dim (:obj:`int`): The embedding dimension of the encoding model (e.g. BERT). This value should \
+            correspond to the model you use. For bert-base-uncased, this value is 768.
```
Comment (PaParaZz1) [inline ding/model/template/language_transformer.py:43]: there should be an indent

File: ding/model/template/language_transformer.py [hunk 10]
```diff
@@ -66,19 +73,27 @@ def _calc_embedding(self, x: list) -> torch.Tensor:
         last_hidden_states = output.hidden_states[-1]
         # Get [CLS] hidden states
         sentence_embedding = last_hidden_states[:, 0, :]  # len(input_list) x hidden_size
+        sentence_embedding = self.norm(sentence_embedding)
```
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply ding/model/template/language_transformer.py:76]: it is `nn.Identity()` in this case

**Review by PaParaZz1 — APPROVED** [review]
